File: scripts/bindings/ggplot_temporal_guides.py
"""FIX-GG04: temporal guide arguments through primary Python authoring and publication."""
import json,math,sys
from pathlib import Path
ROOT=Path(__file__).resolve().parents[2]
sys.path[:0]=[str(ROOT/'packages/python'),str(Path(sys.argv[1]).resolve())]
import finstack_chart as c
out=Path(sys.argv[2]);out.mkdir(parents=True,exist_ok=True)
output=c.Output((ROOT/'fixtures/capability/fonts/NotoSans-Regular.ttf').read_bytes());options=c.export_options(640,360).dpi(96).basis('current')
def descriptor(case,origin,unit,factor):
 control=case['control'];date=case['kind']=='date'
 args={'date':date,'breaks':'Automatic','count':None,'labels':'Automatic','format':None}
 if control=='null_breaks':args['breaks']='None'
 if control=='empty_breaks':args['breaks']={'Explicit':[]}
 if control in ('explicit_breaks','explicit_labels','bad_labels'):args['breaks']={'Explicit':[v*factor for v in (-1,0,2,5)]}
 if control=='width':args['breaks']={'Width':case.get('width', '2 days' if date else '2 secs')}
 if control=='count_two':args['count']=2
 if control=='null_labels':args['labels']='Hidden'
 if control=='explicit_labels':args['labels']={'Explicit':['Before','Start','Two','After']}
 if control=='bad_labels':args['labels']={'Explicit':['A','B']}
 if control=='format':args['format']={'pattern':'%Y-%m-%d' if date else '%H:%M:%S','locale':None}
 limits=[-factor,5*factor] if control=='limits' else [None,5*factor] if control=='partial_limits' else None
 return {'function':{'Interpolated':{'normalization':{'Ggplot':{'family':'Linear','domain':[0,1],'reverse':False,'rescaler':'Range','timestamp':{'origin':str(origin),'unit':unit,'date':date}}},'output':{'Interpolate':{'operation':'PowerRange','range':[1,6],'exponent':0.5,'absolute':False}},'unknown':{'kind':'Missing'}}},'training':'Eligible','ggplot':{'Continuous':{'empty_population':False,'nonfinite_population':False,'limits':limits,'oob':'Censor'}},'guide':'Hidden' if control=='hidden_guide' else {'Temporal':{'origin':str(origin),'unit':unit,'zone':'Utc','arguments':args}}}
records=[]
cases=json.loads((ROOT/'fixtures/parity/ggplot2/temporal-aesthetic-guides.json').read_text())['cases']+json.loads((ROOT/'fixtures/parity/ggplot2/temporal-aesthetic-widths.json').read_text())['cases']
for index,case in enumerate(cases):
 units=[('us','Microseconds',1000000)] if 'width' in case else [('s','Seconds',1),('ms','Milliseconds',1000),('us','Microseconds',1000000),('ns','Nanoseconds',1000000000)]
 for unit,variant,multiplier in units:
  owned=[];expected=case['result'];failure=expected.get('error') or expected.get('draw_error')
  try:
   factor=multiplier*(86400 if case['kind']=='date' else 1);origin=1704067200*multiplier;inputs=case['inputs']
   def stamp(value):
    value=(19723 if case['kind']=='date' else 1704067200) if value is None else value
    whole=math.trunc(value);return whole*factor+round((value-whole)*factor)
   data=c.Data.columns({'x':c.column(list(map(float,range(len(inputs)))),kind='float64'),'y':c.column(list(map(float,range(len(inputs)))),kind='float64'),'when':c.timestamps([stamp(v) for v in inputs],unit,'UTC').validity([v is not None for v in inputs])});owned.append(data)
   scale=descriptor(case,origin,variant,factor);source={'field':'when','origin':str(origin)}
   def layer():return c.points().name('marks').stroke('#000000').numeric_scale('Size',source,scale)
   plot=c.plot(data).profile('Ggplot2_4_0_3').aes(c.aes().x('x').y('y')).layer(layer()).build();owned.append(plot)
   wire=plot.to_json();version=json.loads(wire)['version'];assert version in (25,26)
   restored=c.Plot.from_json(wire);owned.append(restored);assert restored.to_json()==wire
   for state in ('original','edited'):
    current=restored if state=='original' else restored.edit().layer('marks',layer()).build()
    if current is not restored:owned.append(current)
    chart=current.chart();owned.append(chart);actual=chart.semantics()['layers'][0];assert not failure,(index,failure)
    styles=actual.get('styles',[]);wanted=[v for v in expected['values'] if v is not None]
    assert len(styles)==len(wanted),(index,unit,styles,wanted)
    for style,value in zip(styles,wanted):assert abs(style['radius']-value)<2e-12,(index,unit,style,value)
    encoding=actual.get('numeric_scales',{}).get('Size');retained=(encoding or json.loads(current.to_json())['definition']['layers'][0]['numeric_scales']['Size'])['scale'];assert retained['function']['Interpolated']['normalization']['Ggplot']['timestamp']['origin']==str(origin)
    if case['population']=='short' and case['control']=='default':
     standalone=c.StandaloneScale.from_json(json.dumps({'version':5 if case['kind']=='date' else 4,'spec':retained['function']}));owned.append(standalone)
     copied=c.StandaloneScale.from_json(standalone.to_json());owned.append(copied);assert copied.to_json()==standalone.to_json()
     for value,expected_value in zip(inputs,expected['values']):assert abs(copied.map(float(value*factor-origin))-expected_value)<2e-12
    records.append({'index':index,'unit':unit,'state':state,'version':version,'styles':styles,'scale':retained})
    if unit=='s' and state=='original' and ((case['kind'],case['population'],case['control']) in [('date','short','default'),('datetime','short','width'),('date','constant','null_breaks')]):
     request=output.request(current,options);owned.append(request);frame=request.prepare();owned.append(frame)
     for fmt in ('svg','pdf','png'):(out/f"{case['kind']}-{case['population']}-{case['control']}.{fmt}").write_bytes(frame.export(fmt))
  except c.ChartError as error:
   diagnostic=json.loads(str(error));assert failure,(index,unit,diagnostic)
   expected_code='CHART_VALIDATION' if 'width' in case or 'labels' in failure else 'CHART_NUMERICAL_DOMAIN';assert diagnostic['code']==expected_code,(index,diagnostic,failure)
   records.append({'index':index,'unit':unit,'error':diagnostic['code']})
  finally:
   for value in reversed(owned):value.dispose()
assert len(records)==1236,len(records)
(out/'guide-records.json').write_text(json.dumps(records,indent=2));options.dispose();output.dispose();print('PASS Python: 1236 temporal guide states and nine publication files.')

File: crates/chart-core/tests/ggplot_temporal_guides.rs
//! FIX-GG04: reference Date/datetime aesthetic guide arguments and population precedence.
use chart_core::{
    data::TimeUnit,
    grammar::NumericAesthetic,
    interpolate::{Number, Value},
    prelude::*,
    scales::*,
};
use serde_json::Value as Json;
fn spec(
    case: &Json,
    origin: i64,
    unit: TimeUnit,
    factor: f64,
) -> chart_core::ChartResult<MappedScaleSpec> {
    let mut scale = ggplot_numeric_default(GgplotNumericPalette::Size)?
        .with_timestamp_normalization(GgplotTimestampNormalization {
            origin,
            unit,
            date: case["kind"] == "date",
        })?;
    let control = case["control"].as_str().unwrap();
    if let Some(GgplotScalePolicy::Continuous { limits, .. }) = scale.ggplot.as_deref_mut() {
        if control == "limits" {
            *limits = Some([Some(Number(-factor)), Some(Number(5. * factor))]);
        }
        if control == "partial_limits" {
            *limits = Some([None, Some(Number(5. * factor))]);
        }
    }
    let mut args = GgplotTemporalGuideArguments {
        date: case["kind"] == "date",
        ..Default::default()
    };
    match control {
        "null_breaks" => args.breaks = GgplotTemporalBreaks::None,
        "empty_breaks" => args.breaks = GgplotTemporalBreaks::Explicit(vec![]),
        "explicit_breaks" | "explicit_labels" | "bad_labels" => {
            args.breaks = GgplotTemporalBreaks::Explicit(
                [-1., 0., 2., 5.].map(|v| Number(v * factor)).to_vec(),
            )
        }
        "width" => {
            args.breaks = GgplotTemporalBreaks::Width(
                case["width"]
                    .as_str()
                    .unwrap_or(if args.date { "2 days" } else { "2 secs" })
                    .into(),
            )
        }
        "count_two" => args.count = Some(2.),
        _ => (),
    }
    match control {
        "null_labels" => args.labels = GgplotGuideLabels::Hidden,
        "explicit_labels" => {
            args.labels = GgplotGuideLabels::Explicit(
                ["Before", "Start", "Two", "After"]
                    .map(|v| Some(v.into()))
                    .to_vec(),
            )
        }
        "bad_labels" => {
            args.labels = GgplotGuideLabels::Explicit(["A", "B"].map(|v| Some(v.into())).to_vec())
        }
        "format" => {
            args.format = Some(Box::new(GgplotTimeFormat {
                pattern: if args.date { "%Y-%m-%d" } else { "%H:%M:%S" }.into(),
                locale: None,
            }))
        }
        _ => (),
    }
    scale = scale.with_guide(GgplotScaleGuide::Temporal(GgplotTemporalGuide {
        origin,
        unit,
        zone: CalendarZone::Utc,
        arguments: args,
    }))?;
    if control == "hidden_guide" {
        scale = scale.with_guide(GgplotScaleGuide::Hidden)?;
    }
    Ok(scale)
}
#[test]
fn temporal_arguments_match_reference_mapping_guides_json_and_edits() {
    let fixture: Json = serde_json::from_str(include_str!(
        "../../../fixtures/parity/ggplot2/temporal-aesthetic-guides.json"
    ))
    .unwrap();
    assert_eq!(fixture["cases"].as_array().unwrap().len(), 156);
    run_cases(
        fixture["cases"].as_array().unwrap(),
        &[
            (TimeUnit::Seconds, 1),
            (TimeUnit::Milliseconds, 1000),
            (TimeUnit::Microseconds, 1000000),
            (TimeUnit::Nanoseconds, 1000000000),
        ],
    );
}
#[test]
fn temporal_width_units_fractional_alignment_and_rejections_match_reference() {
    let fixture: Json = serde_json::from_str(include_str!(
        "../../../fixtures/parity/ggplot2/temporal-aesthetic-widths.json"
    ))
    .unwrap();
    assert_eq!(fixture["cases"].as_array().unwrap().len(), 60);
    // Microseconds cover fractional seconds and multi-year spans with exact offsets.
    run_cases(
        fixture["cases"].as_array().unwrap(),
        &[(TimeUnit::Microseconds, 1000000)],
    );
}
fn run_cases(cases: &[Json], units: &[(TimeUnit, i64)]) {
    for case in cases {
        for &(unit, multiplier) in units {
            let date = case["kind"] == "date";
            let factor = multiplier * if date { 86400 } else { 1 };
            let origin = 1704067200 * multiplier;
            let inputs = case["inputs"].as_array().unwrap();
            let stamps: Vec<_> = inputs
                .iter()
                .map(|v| {
                    let value = v
                        .as_f64()
                        .unwrap_or(if date { 19723. } else { 1704067200. });
                    let whole = value.trunc() as i64;
                    whole * factor + ((value - whole as f64) * factor as f64).round() as i64
                })
                .collect();
            let data = Data::columns()
                .column("x", (0..inputs.len()).map(|i| i as f64).collect::<Vec<_>>())
                .column("y", (0..inputs.len()).map(|i| i as f64).collect::<Vec<_>>())
                .column(
                    "when",
                    timestamps(stamps.clone(), unit, "UTC")
                        .validity(inputs.iter().map(|v| !v.is_null()).collect()),
                )
                .build()
                .unwrap();
            let figure = spec(case, origin, unit, factor as f64).and_then(|scale| {
                plot(data)
                    .profile(Profile::Ggplot2_4_0_3)
                    .aes(aes().x("x").y("y"))
                    .layer(points().name("marks").numeric_scale(
                        NumericAesthetic::Size,
                        Mapping::Timestamp {
                            field: "when".into(),
                            origin,
                        },
                        scale,
                    ))
                    .build()
            });
            let expected = &case["result"];
            if expected["error"].is_string() || expected["draw_error"].is_string() {
                assert!(
                    figure.and_then(|p| p.chart()?.prepare()).is_err(),
                    "expected failure: {case} {unit:?}"
                );
                continue;
            }
            let figure = figure.unwrap_or_else(|e| panic!("{case} {unit:?}: {e:?}"));
            let figure = Plot::from_json(&figure.to_json().unwrap()).unwrap();
            for current in [
                figure.clone(),
                figure
                    .edit()
                    .layer(
                        "marks",
                        points().stroke(rgb(1, 2, 3)).numeric_scale(
                            NumericAesthetic::Size,
                            Mapping::Timestamp {
                                field: "when".into(),
                                origin,
                            },
                            spec(case, origin, unit, factor as f64).unwrap(),
                        ),
                    )
                    .build()
                    .unwrap(),
            ] {
                let prepared = current
                    .chart()
                    .unwrap()
                    .prepare()
                    .unwrap_or_else(|e| panic!("{case} {unit:?}: {e:?}"));
                let layer = &prepared.layers()[0];
                let encoding = layer
                    .numeric_scales()
                    .get(&NumericAesthetic::Size)
                    .unwrap_or_else(|| panic!("missing scale: {case} {unit:?}"));
                let scale = MappedScale::for_numbers(encoding.scale.clone()).unwrap();
                let entries = scale.continuous_guide_entries(128, 65536).unwrap();
                let entries = entries.unwrap();
                if case["control"] == "hidden_guide"
                    || case["population"] == "empty" && case["control"] != "limits"
                {
                    assert!(entries.is_empty(), "{case}");
                } else {
                    let breaks = expected["guide"]["breaks"]
                        .as_array()
                        .unwrap_or_else(|| panic!("{case}"));
                    assert_eq!(entries.len(), breaks.len(), "{case} {unit:?}: {entries:?}");
                    for (index, (entry, expected_break)) in entries.iter().zip(breaks).enumerate() {
                        let absolute =
                            (i128::from(origin) + entry.value.0 as i128) as f64 / factor as f64;
                        assert!(
                            (absolute - expected_break.as_f64().unwrap()).abs() < 1e-8,
                            "{case}: {entry:?}"
                        );
                        let label = expected["guide"]["labels"]
                            .get(index)
                            .and_then(Json::as_str);
                        assert_eq!(entry.label.as_deref(), label, "{case} {unit:?}");
                    }
                }
                for ((stamp, input), expected_value) in stamps
                    .iter()
                    .zip(inputs)
                    .zip(expected["values"].as_array().unwrap())
                {
                    let value = scale
                        .numeric(if input.is_null() {
                            None
                        } else {
                            Some((i128::from(*stamp) - i128::from(origin)) as f64)
                        })
                        .unwrap();
                    if let Some(expected_value) = expected_value.as_f64() {
                        let Value::Number(actual) = value else {
                            panic!("{case}: {value:?}")
                        };
                        assert!(
                            (actual.0 - expected_value).abs() < 2e-12,
                            "{case}: {actual:?}"
                        );
                    } else {
                        assert!(
                            matches!(value, Value::Missing | Value::Null),
                            "{case}: {value:?}"
                        );
                    }
                }
            }
        }
    }
}
